### PR TITLE
made the changes to remove the old compatibility layer of NDData

### PR DESCRIPTION
**Pull request by abit2**
https://github.com/sunpy/sunpy/issues/1302 changes made to support the latest NDData and made a methods that returns a new map object with the modified data.

### CODE REVIEW

File: sunpy/map/nddata_compat.py [hunk 1]
```diff
@@ -14,94 +14,81 @@
```
Comment (Cadair) [inline sunpy/map/nddata_compat.py]: why does this file still exist?!

File: sunpy/map/mapbase.py [hunk 2]
```diff
@@ -221,6 +221,9 @@ def __repr__(self):
            dim=u.Quantity(self.dimensions),
            scale=u.Quantity(self.scale),
            tmf=TIME_FORMAT) + self.data.__repr__())
+                                                                        
+    def _modified_copy(self, data, meta, **kwargs):
+        return GenericMap(data, meta, **kwargs)
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: I would persist `plot_settings` here as well (i.e copy plot settings from `self` to the new map).

File: sunpy/map/mapbase.py [hunk 3]
```diff
@@ -221,6 +221,9 @@ def __repr__(self):
            dim=u.Quantity(self.dimensions),
            scale=u.Quantity(self.scale),
            tmf=TIME_FORMAT) + self.data.__repr__())
+                                                                        
+    def _modified_copy(self, data, meta, **kwargs):
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: This should also have a docstring for developers to understand what's going on.

File: sunpy/map/nddata_compat.py (hunk 1 above)
Comment (abit2) [inline sunpy/map/nddata_compat.py]: i thought you have made a few extra conditional check on uncertainity and mask maybe

**Comment by abit2** [comment]
https://travis-ci.org/sunpy/sunpy/jobs/169592510#L1149 @Cadair how do i solve this ?
Checked the meta data before and after creating modified new map it's same, didn't understand please help

File: sunpy/map/mapbase.py [hunk 4]
```diff
@@ -221,6 +221,12 @@ def __repr__(self):
            dim=u.Quantity(self.dimensions),
            scale=u.Quantity(self.scale),
            tmf=TIME_FORMAT) + self.data.__repr__())
+                                                                        
+    def _modified_copy(self, data, meta, plot_settings=None, **kwargs):
+        """
+        Method to create new modified map object with the required changes
+        """
+        return GenericMap(data, meta, **kwargs)
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: This is the root of the test failure. You need to return the same type as `self` not just a `GenericMap` to fix the test you need to replace this by:

```python
new_map = type(self)(data, meta, **kwargs)
return new_map
```

Also you need to set plot settings in this function:

```python
new_map = type(self)(data, meta, **kwargs)
if plot_settings:
    new_map.plot_settings = plot_settings
return new_map
```

File: sunpy/map/mapbase.py [hunk 5]
```diff
@@ -962,44 +969,44 @@ def rotate(self, angle=None, rmatrix=None, order=4, scale=1.0,
...
+        pixel_rotation_center = u.Quantity(self.data_to_pixel(*rotation_center,
                                                                  origin=0)).value
         if recenter:
             pixel_center = pixel_rotation_center
         else:
             pixel_center = pixel_array_center
 
         # Apply the rotation to the image data
-        new_map.data = affine_transform(new_map.data.T,
+        new_data = affine_transform(new_data.T,
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: the following lines here need their indentation modified for style.

**Comment by ayshih** [comment]
Per discussion during today's dev meeting, `_modified_copy()` is confusingly named, @Cadair learned that class methods can be called on instances, and there's probably a cleaner implementation by expanding `__init__()`.

**Comment by abit2** [comment]
@ayshih can you please elaborate on the expanding __init__() and what should be a better name?

**Comment by Cadair** [comment]
@ayshih So the evil plan does not work
screenshot from 2016-11-02 13-21-28

@abit2 I am still working out the implications of the discussion at the dev meeting. I will give you more info or maybe a PR.

**Comment by Cadair** [comment]
@ayshih basically I came to the conclusion that the best option was a simple class method.

**Comment by ayshih** [comment]
Oh well.  Now I know that even though `__new__(cls, ...)` looks like a class method, it's actually a static method as a special case.  That is, `cls` needs to be explicitly passed in the call, which is why the "evil plan" doesn't work!

**Comment by abit2** [comment]
@Cadair having problem resolving the conflict, when merging upstream/master frames.py is also getting added

**Comment by abit2** [comment]
@Cadair @dpshelio 
gitmate??

**Comment by dpshelio** [comment]
ack 06baddc

**Comment by abit2** [comment]
@dpshelio Sorry didn't understand "ack 06baddc"

**Comment by abit2** [comment]
@dpshelio Sorry didn't understand ack 06baddc

**Pull request merged**